models/demos/mamba/demo/demo.py

```diff
@@ -0,0 +1,152 @@
+    from models.demos.mamba.reference.decode_model import MambaDecode
+
+    return MambaDecode.from_pretrained(version, batch_size=batch_size)
+
+
+def get_tt_metal_model(version: MambaPretrainedModelName, device: ttnn.Device, use_cache: bool, batch_size: int):
+    from models.demos.mamba.tt.full_model import MambaTT
+    from models.demos.mamba.tt import model_config
+
+    reference_model = get_cpu_reference_model(version, batch_size=batch_size)
+    if use_cache:
+        cache_path = f"/tmp/{version}"
+    else:
+        cache_path = None
+
+    config = model_config.create_model_config(batch_size, reference_model.args.d_model)
+    model = MambaTT(reference_model, device, config, tt_cache_path=cache_path, num_layers=64)
+
+    return model
+
+
+def get_tokenizer():
+    tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-neox-20b", padding=True)
+    tokenizer.pad_token_id = tokenizer.eos_token_id
+    return tokenizer
+
+
+def display_tokens(tokens: List[str]):
+    print("\n" * 1000)
+    for text in tokens:
+        print(f"{text}\n")
+        print("-" * 150)  # Print a separator line for readability
+        print(f"\n")
+
+
+def apply_repetition_penalty_(logits, sequences, penalty=1.2):
+    """
+    Applies a repetition penalty to the logits.
+
+    Args:
+        logits (torch.Tensor): The logits from the model, shape [batch_size, vocab_size].
+        sequences (torch.Tensor): The generated sequences so far, shape [batch_size, sequence_length].
+        penalty (float): The penalty factor for repeated tokens (values > 1 discourage repetition).
+    """
+    batch_size, vocab_size = logits.shape
+    for i in range(batch_size):
+        for j in range(sequences.shape[1]):
+            token_id = sequences[i, j]
+            if token_id < vocab_size:  # Protect against potential out-of-bounds errors
+                logits[i, token_id] /= penalty  # Decrease the probability of already generated tokens
+    return logits
+
+
+def run_demo(
+    prompts: List[str],
+    model_type: str,
+    model_version: MambaPretrainedModelName = "state-spaces/mamba-2.8b-slimpj",
+    generated_sequence_length: int = 50,
+    use_cache: bool = True,
+):
+    batch_size = len(prompts)
+    assert batch_size == 32, "32 prompts are required"
+    device = None
+
+    logger.info(f"Running Mamba demo (weights='{model_version}') on '{model_type.upper()}' with batch={batch_size}")
+
+    if model_type == "cpu":
+        model = get_cpu_reference_model(model_version, batch_size)
+    elif model_type == "wh":
+        device = ttnn.open_device(device_id=0)
+        ttnn.enable_program_cache(device)
+        model = get_tt_metal_model(model_version, device, use_cache, batch_size)
+    else:
+        raise RuntimeError("Invalid model type was encountered")
+
+    tokenizer = get_tokenizer()
+
+    sequences: torch.Tensor = tokenizer(prompts, return_tensors="pt", padding=True).input_ids
+
+    # prefill
+    prefill_iterations = sequences.shape[1] - 1
```

Comment:
I am wondering if we should prefill on CPU and push the resulting hidden state to silicon. This will make the demo faster and help focus on what is important.

Reply:
Pushing this work to next PR